src/lib/UI/Module/ContentTree/NodeFactory.php

```diff
@@ -0,0 +1,214 @@
+use eZ\Publish\API\Repository\Values\Content\LocationQuery;
+use eZ\Publish\API\Repository\Values\Content\Query\Criterion;
+use eZ\Publish\API\Repository\Values\Content\Search\SearchResult;
+use EzSystems\EzPlatformAdminUi\REST\Value\ContentTree\LoadSubtreeRequestNode;
+use EzSystems\EzPlatformAdminUi\REST\Value\ContentTree\Node;
+
+/**
+ * @internal
+ */
+final class NodeFactory
+{
+    /** @var \eZ\Publish\API\Repository\SearchService */
+    private $searchService;
+
+    /** @var int */
+    private $displayLimit;
+
+    /** @var int */
+    private $childrenLoadMaxLimit;
+
+    /** @var int */
+    private $maxDepth;
+
+    /** @var string[] */
+    private $contentTypeIdentifierIgnoreList;
+
+    /**
+     * @param \eZ\Publish\API\Repository\SearchService $searchService
+     * @param int $displayLimit
+     * @param int $childrenLoadMaxLimit
+     * @param int $maxDepth
+     * @param array $contentTypeIdentifierIgnoreList
+     */
+    public function __construct(
+        SearchService $searchService,
+        int $displayLimit = 20,
+        int $childrenLoadMaxLimit = 100,
+        int $maxDepth = 10,
+        array $contentTypeIdentifierIgnoreList = []
+    ) {
+        $this->searchService = $searchService;
+        $this->displayLimit = $displayLimit;
+        $this->childrenLoadMaxLimit = $childrenLoadMaxLimit;
+        $this->maxDepth = $maxDepth;
+        $this->contentTypeIdentifierIgnoreList = $contentTypeIdentifierIgnoreList;
+    }
+
+    /**
+     * @param \eZ\Publish\API\Repository\Values\Content\Location $location
+     * @param \EzSystems\EzPlatformAdminUi\REST\Value\ContentTree\LoadSubtreeRequestNode $loadSubtreeRequestNode
+     * @param bool $loadChildren
+     * @param int $depth
+     *
+     * @return \EzSystems\EzPlatformAdminUi\REST\Value\ContentTree\Node
+     *
+     * @throws \eZ\Publish\API\Repository\Exceptions\InvalidArgumentException
+     * @throws \eZ\Publish\API\Repository\Exceptions\NotFoundException
+     * @throws \eZ\Publish\API\Repository\Exceptions\UnauthorizedException
+     */
+    public function createNode(
+        Location $location,
+        ?LoadSubtreeRequestNode $loadSubtreeRequestNode = null,
+        bool $loadChildren = false,
+        int $depth = 0
+    ): Node {
+        $content = $location->getContent();
+        $contentType = $content->getContentType();
+
+        $numberOfSubitemsToLoad = $this->resolveLoadLimit($loadSubtreeRequestNode);
+
+        $children = [];
+        if ($depth < $this->maxDepth && $loadChildren) {
+            $searchResult = $this->findSubitems($location, $numberOfSubitemsToLoad);
+            $totalChildrenCount = $searchResult->totalCount;
+
+            /** @var \eZ\Publish\API\Repository\Values\Content\Location $childLocation */
+            foreach (array_column($searchResult->searchHits, 'valueObject') as $childLocation) {
+                $childLoadSubtreeRequestNode = null !== $loadSubtreeRequestNode
```

Comment:
IMHO this expression is complex enough to extract it to separate method (might replace `findChild` which is used only here).   

```php
private function resolveChildLoadSubtreeRequestNode(?LoadSubtreeRequestNode $loadSubtreeRequestNode, Location $location): ?LoadSubtreeRequestNode
{
    if ($loadSubtreeRequestNode !== null) {
        foreach ($loadSubtreeRequestNode->children as $child) {
            if ($child->locationId === $location->id) {
                return $child;
            }
        }
    }

    return null;
}
```

Reply:
Might be a matter of personal preference but to me it makes little sense to allow nullable argument and ignore the method's logic if it's null. To me it's better to have it outside of the method. Any opinion on that?